compiler/rustc_mir/src/transform/const_goto.rs

```diff
@@ -0,0 +1,129 @@
+
+use super::simplify::{simplify_cfg, simplify_locals};
+
+pub struct ConstGoto;
+
+impl<'tcx> MirPass<'tcx> for ConstGoto {
+    fn run_pass(&self, tcx: TyCtxt<'tcx>, source: MirSource<'tcx>, body: &mut Body<'tcx>) {
+        trace!("Running ConstGoto on {:?}", source);
+        let param_env = tcx.param_env_reveal_all_normalized(source.def_id());
+        let mut opt_finder =
+            ConstGotoOptimizationFinder { tcx, body, optimizations: vec![], param_env };
+        opt_finder.visit_body(body);
+        let should_simplify = !opt_finder.optimizations.is_empty();
+        for opt in opt_finder.optimizations {
+            let terminator = body.basic_blocks_mut()[opt.bb_with_goto].terminator_mut();
+            let new_goto = TerminatorKind::Goto { target: opt.target_to_use_in_goto };
+            debug!("SUCCESS: replacing `{:?}` with `{:?}`", terminator.kind, new_goto);
+            terminator.kind = new_goto;
+        }
+
+        // if we applied optimizations, we potentially have some cfg to cleanup to
+        // make it easier for further passes
+        if should_simplify {
+            simplify_cfg(body);
+            simplify_locals(body, tcx);
+        }
+    }
+}
+
+impl<'a, 'tcx> Visitor<'tcx> for ConstGotoOptimizationFinder<'a, 'tcx> {
+    fn visit_terminator(&mut self, terminator: &Terminator<'tcx>, location: Location) {
+        let mut bailer = || {
+            match terminator.kind {
+                TerminatorKind::Goto { target } => {
+                    // We only apply this optimization if the last statement is a const assignment
+                    let last_statement =
+                        self.body.basic_blocks()[location.block].statements.last()?;
+
+                    match &last_statement.kind {
+                        StatementKind::Assign(box (place, Rvalue::Use(op))) => {
+                            let _const = op.constant()?;
+                            // We found a constant being assigned to `place`.
+                            // Now check that the target of this Goto switches on this place.
+                            let target_bb = &self.body.basic_blocks()[target];
+                            if !target_bb.statements.is_empty() {
+                                return None;
+                            }
+
+                            let target_bb_terminator = target_bb.terminator();
+                            match &target_bb_terminator.kind {
+                                TerminatorKind::SwitchInt { discr, switch_ty, values, targets }
+                                    if discr.place() == Some(*place) =>
+                                {
+                                    // We now know that the Switch matches on the const place, and it is statementless
+                                    // Now find which value in the Switch matches the const value.
+                                    let const_value = _const.literal.eval_bits(
```

Comment:
Const evaluation might fail (there is try_eval_bits):

```rust
#![feature(min_const_generics)]
#![crate_type = "lib"]

pub fn f<const A: i32, const B: bool>() -> u64 {
    match { 
        match A {
            1 | 2 | 3 => B,
            _ => true 
        }
    } {
      false => 1,
      true => 2,
    }
}
```

Reply:
Thanks, fixed on latest commit.